pkg/query-service/app/logs/v3/query_builder.go

```diff
@@ -0,0 +1,326 @@
+	v3.AggregateOperatorHistQuant75: 0.75,
+	v3.AggregateOperatorHistQuant90: 0.90,
+	v3.AggregateOperatorHistQuant95: 0.95,
+	v3.AggregateOperatorHistQuant99: 0.99,
+}
+
+var aggregateOperatorToSQLFunc = map[v3.AggregateOperator]string{
+	v3.AggregateOperatorAvg:     "avg",
+	v3.AggregateOperatorMax:     "max",
+	v3.AggregateOperatorMin:     "min",
+	v3.AggregateOperatorSum:     "sum",
+	v3.AggregateOperatorRateSum: "sum",
+	v3.AggregateOperatorRateAvg: "avg",
+	v3.AggregateOperatorRateMax: "max",
+	v3.AggregateOperatorRateMin: "min",
+}
+
+var logsOperatorMappingV3 = map[string]string{
+	"eq":     "=",
+	"neq":    "!=",
+	"lt":     "<",
+	"lte":    "<=",
+	"gt":     ">",
+	"gte":    ">=",
+	"like":   "LIKE",
+	"nlike":  "NOT LIKE",
+	"regex":  "REGEXP",
+	"nregex": "NOT REGEXP",
+	"in":     "IN",
+	"nin":    "NOT IN",
+	// (todo) check contains/not contains/exists/not exists
+}
+
+// getClickhouseColumnName returns the corresponding clickhouse column name for the given attribute/resource key
+func getClickhouseColumnName(key v3.AttributeKey, fields map[string]v3.AttributeKey) (string, error) {
+	clickhouseColumn := key.Key
+	//if the key is present in the topLevelColumn then it will be only searched in those columns,
+	//regardless if it is indexed/present again in resource or column attribute
+	_, isTopLevelCol := constants.LogsTopLevelColumnsV3[key.Key]
+	if !isTopLevelCol && !key.IsColumn {
+		// check if we need to enrich metadata
+		if key.Type == "" || key.DataType == "" {
+			// check if the field is present in the fields map
+			if field, ok := fields[key.Key]; ok {
+				if field.IsColumn {
+					return field.Key, nil
+				}
+				key.Type = field.Type
+				key.DataType = field.DataType
+			} else {
+				return "", fmt.Errorf("field not found to enrich metadata")
+			}
+		}
+
+		columnType := key.Type
+		if columnType == v3.AttributeKeyTypeTag {
+			columnType = "attributes"
+		} else {
+			columnType = "resources"
+		}
+
+		columnDataType := "string"
+		if key.DataType == v3.AttributeKeyDataTypeFloat64 {
+			columnDataType = "float64"
+		}
+		if key.DataType == v3.AttributeKeyDataTypeInt64 {
+			columnDataType = "int64"
+		}
+		clickhouseColumn = fmt.Sprintf("%s_%s_value[indexOf(%s_%s_key, '%s')]", columnType, columnDataType, columnType, columnDataType, key.Key)
+	}
+	return clickhouseColumn, nil
+}
+
+// getSelectLabels returns the select labels for the query based on groupBy and aggregateOperator
+func getSelectLabels(aggregatorOperator v3.AggregateOperator, groupBy []v3.AttributeKey, fields map[string]v3.AttributeKey) (string, error) {
+	var selectLabels string
+	if aggregatorOperator == v3.AggregateOperatorNoOp || aggregatorOperator == v3.AggregateOperatorRate {
+		selectLabels = ""
+	} else {
+		for _, tag := range groupBy {
+			columnName, err := getClickhouseColumnName(tag, fields)
+			if err != nil {
+				return "", err
+			}
+			selectLabels += fmt.Sprintf(", %s as %s", columnName, tag.Key)
+		}
+	}
```

Comment:
The `rate` in metrics is a bit different from the `rate` in logs, You could have the rate of logs by service, host, so you need to handle that. You seem to skip that for `rate` here?